File: Ascentix.RulesEngine.Core/Execution/MathExpr.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using Microsoft.Xrm.Sdk;

namespace Ascentix.RulesEngine.Core.Execution
{
    /// <summary>AST for a numeric field-mapping expression (source == "mathexpr").</summary>
    public abstract class MathExprNode { }
    public sealed class NumberNode : MathExprNode { public decimal Value; }
    public sealed class ColumnRefNode : MathExprNode { public Guid? Node; public string Column; }
    public sealed class UnaryNode : MathExprNode { public MathExprNode Operand; } // arithmetic negate
    public sealed class BinaryNode : MathExprNode { public char Op; public MathExprNode Left, Right; }
    public enum AggregateFunc { Sum, Avg, Min, Max, Count }
    public sealed class AggregateNode : MathExprNode { public AggregateFunc Func; public Guid Node; public string Column; public string FilterKey; } // Column null for Count; FilterKey null when unfiltered

    /// <summary>
    /// Recursive-descent parser for a numeric expression: numeric literals, {root.col} /
    /// {node:guid.col} tokens (same grammar as TemplateRenderer), aggregate operands
    /// (sum/avg/min/max/count), binary + - * /, unary -, and parentheses with standard
    /// precedence (* / over + -), left-associative. Pure and sandbox-safe (no eval).
    /// A malformed expression throws InvalidPluginExecutionException, prefixed with errorContext.
    /// </summary>
    public static class MathExpr
    {
        public static MathExprNode Parse(string expression, string errorContext)
        {
            var tokens = Tokenize(expression ?? "", errorContext);
            var pos = 0;
            var node = ParseExpr(tokens, ref pos, errorContext);
            if (pos != tokens.Count)
                throw Err(errorContext, "unexpected trailing input in expression.");
            return node;
        }

        // The three reference walks below are implementation details of RuleReferences (the
        // rule reference set); consumers read its named answers rather than walking an AST.
        internal static IEnumerable<(Guid? node, string column)> ExtractRefs(MathExprNode ast)
        {
            switch (ast)
            {
                case ColumnRefNode c: yield return (c.Node, c.Column); break;
                case AggregateNode a: yield return (a.Node, a.Column); break;
                case UnaryNode u:
                    foreach (var r in ExtractRefs(u.Operand)) yield return r; break;
                case BinaryNode b:
                    foreach (var r in ExtractRefs(b.Left)) yield return r;
                    foreach (var r in ExtractRefs(b.Right)) yield return r;
                    break;
            }
        }

        /// <summary>Scalar column operands ({root.col} / {node:guid.col}) only: must resolve to a
        /// single record, so any node they reference must be single-cardinality. Excludes aggregates.</summary>
        internal static IEnumerable<(Guid? node, string column)> ScalarRefs(MathExprNode ast)
        {
            switch (ast)
            {
                case ColumnRefNode c: yield return (c.Node, c.Column); break;
                case UnaryNode u:
                    foreach (var r in ScalarRefs(u.Operand)) yield return r; break;
                case BinaryNode b:
                    foreach (var r in ScalarRefs(b.Left)) yield return r;
                    foreach (var r in ScalarRefs(b.Right)) yield return r;
                    break;
            }
        }

        /// <summary>Aggregate operands (sum/avg/min/max/count(node:...)) only: reduce a child
        /// collection, so the node they reference must be many-cardinality.</summary>
        internal static IEnumerable<(Guid node, string column)> AggregateRefs(MathExprNode ast)
        {
            switch (ast)
            {
                case AggregateNode a: yield return (a.Node, a.Column); break;
                case UnaryNode u:
                    foreach (var r in AggregateRefs(u.Operand)) yield return r; break;
                case BinaryNode b:
                    foreach (var r in AggregateRefs(b.Left)) yield return r;
                    foreach (var r in AggregateRefs(b.Right)) yield return r;
                    break;
            }
        }

        /// <summary>Every <see cref="AggregateNode"/> in the tree (not just its node/column), so
        /// callers can inspect <see cref="AggregateNode.FilterKey"/> (e.g. field-mapping filter
        /// key-integrity checks).</summary>
        public static IEnumerable<AggregateNode> AggregateNodes(MathExprNode ast)
        {
            switch (ast)
            {
                case AggregateNode a: yield return a; break;
                case UnaryNode u:
                    foreach (var r in AggregateNodes(u.Operand)) yield return r; break;
                case BinaryNode b:
                    foreach (var r in AggregateNodes(b.Left)) yield return r;
                    foreach (var r in AggregateNodes(b.Right)) yield return r;
                    break;
            }
        }

        // ── token stream ────────────────────────────────────────────────
        private enum TokKind { Number, Ref, Agg, Plus, Minus, Star, Slash, LParen, RParen }
        private struct Tok { public TokKind Kind; public decimal Num; public Guid? Node; public string Col; public AggregateFunc Func; public string FilterKey; }

        private static List<Tok> Tokenize(string s, string ctx)
        {
            var toks = new List<Tok>();
            var i = 0;
            while (i < s.Length)
            {
                var c = s[i];
                if (char.IsWhiteSpace(c)) { i++; continue; }
                switch (c)
                {
                    case '+': toks.Add(new Tok { Kind = TokKind.Plus }); i++; continue;
                    case '-': toks.Add(new Tok { Kind = TokKind.Minus }); i++; continue;
                    case '*': toks.Add(new Tok { Kind = TokKind.Star }); i++; continue;
                    case '/': toks.Add(new Tok { Kind = TokKind.Slash }); i++; continue;
                    case '(': toks.Add(new Tok { Kind = TokKind.LParen }); i++; continue;
                    case ')': toks.Add(new Tok { Kind = TokKind.RParen }); i++; continue;
                }
                if (c == '{')
                {
                    var close = s.IndexOf('}', i + 1);
                    if (close < 0) throw Err(ctx, "expression has an unclosed '{' token.");
                    var (node, col) = ParseToken(s.Substring(i + 1, close - i - 1), ctx);
                    toks.Add(new Tok { Kind = TokKind.Ref, Node = node, Col = col });
                    i = close + 1; continue;
                }
                if (char.IsDigit(c) || c == '.')
                {
                    var start = i;
                    while (i < s.Length && (char.IsDigit(s[i]) || s[i] == '.')) i++;
                    var lit = s.Substring(start, i - start);
                    if (!decimal.TryParse(lit, NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out var num))
                        throw Err(ctx, $"'{lit}' is not a valid number.");
                    toks.Add(new Tok { Kind = TokKind.Number, Num = num }); continue;
                }
                if (char.IsLetter(c))
                {
                    var start = i;
                    while (i < s.Length && char.IsLetter(s[i])) i++;
                    var name = s.Substring(start, i - start).ToLowerInvariant();
                    var func = ParseFunc(name, ctx);
                    while (i < s.Length && char.IsWhiteSpace(s[i])) i++;
                    if (i >= s.Length || s[i] != '(') throw Err(ctx, $"'{name}' must be followed by '('.");
                    var close = s.IndexOf(')', i + 1);
                    if (close < 0) throw Err(ctx, $"'{name}(' has no closing ')'.");
                    var (node, col, filterKey) = ParseAggArg(s.Substring(i + 1, close - i - 1).Trim(), func, name, ctx);
                    toks.Add(new Tok { Kind = TokKind.Agg, Func = func, Node = node, Col = col, FilterKey = filterKey });
                    i = close + 1; continue;
                }
                throw Err(ctx, $"unexpected character '{c}' in expression.");
            }
            return toks;
        }

        // {root.<col>} or {node:<guid>.<col>}. Mirrors TemplateRenderer.ParseToken.
        private static (Guid? node, string col) ParseToken(string token, string ctx)
        {
            if (token.StartsWith("root.", StringComparison.Ordinal))
            {
                var col = token.Substring("root.".Length);
                if (col.Length == 0) throw Err(ctx, $"token '{{{token}}}' is missing a column name.");
                return (null, col);
            }
            if (token.StartsWith("node:", StringComparison.Ordinal))
            {
                var rest = token.Substring("node:".Length);
                var dot = rest.IndexOf('.');
                if (dot <= 0 || dot == rest.Length - 1)
                    throw Err(ctx, $"token '{{{token}}}' must be '{{node:<guid>.<column>}}'.");
                if (!Guid.TryParse(rest.Substring(0, dot), out var id))
                    throw Err(ctx, $"token '{{{token}}}' does not contain a valid node GUID.");
                return (id, rest.Substring(dot + 1));
            }
            throw Err(ctx, $"unknown token '{{{token}}}'. Expected '{{root.<column>}}' or '{{node:<guid>.<column>}}'.");
        }

        private static AggregateFunc ParseFunc(string name, string ctx)
        {
            switch (name)
            {
                case "sum": return AggregateFunc.Sum;
                case "avg": return AggregateFunc.Avg;
                case "min": return AggregateFunc.Min;
                case "max": return AggregateFunc.Max;
                case "count": return AggregateFunc.Count;
                default: throw Err(ctx, $"unknown function '{name}'. Expected sum, avg, min, max, or count.");
            }
        }

        // Aggregate argument is a brace-free 'node:<guid>[.<column>] [filter:<key>]'. Column required except for count.
        private static (Guid node, string col, string filterKey) ParseAggArg(string arg, AggregateFunc func, string name, string ctx)
        {
            string filterKey = null;
            var ws = arg.IndexOfAny(new[] { ' ', '\t' });
            if (ws >= 0)
            {
                var tail = arg.Substring(ws + 1).Trim();
                arg = arg.Substring(0, ws);
                if (!tail.StartsWith("filter:", StringComparison.Ordinal))
                    throw Err(ctx, $"{name}(...) has unexpected text '{tail}'; expected 'filter:<key>'.");
                filterKey = tail.Substring("filter:".Length).Trim();
                if (filterKey.Length == 0 || !IsFilterKey(filterKey))
                    throw Err(ctx, $"{name}(...) filter key '{filterKey}' must be one or more letters, digits, or underscores.");
            }
            if (!arg.StartsWith("node:", StringComparison.Ordinal))
                throw Err(ctx, $"{name}(...) argument must be 'node:<guid>{(func == AggregateFunc.Count ? "" : ".<column>")}'.");
            var rest = arg.Substring("node:".Length);
            var dot = rest.IndexOf('.');
            var guidPart = dot < 0 ? rest : rest.Substring(0, dot);
            var col = dot < 0 ? null : rest.Substring(dot + 1);
            if (!Guid.TryParse(guidPart, out var id))
                throw Err(ctx, $"{name}(...) does not contain a valid node GUID.");
            if (func == AggregateFunc.Count)
            {
                if (!string.IsNullOrEmpty(col)) throw Err(ctx, "count(...) takes no column. Use 'count(node:<guid>)'.");
            }
            else if (string.IsNullOrEmpty(col))
            {
                throw Err(ctx, $"{name}(...) needs a column. Use '{name}(node:<guid>.<column>)'.");
            }
            return (id, col, filterKey);
        }

        private static bool IsFilterKey(string s)
        {
            foreach (var c in s) if (!char.IsLetterOrDigit(c) && c != '_') return false;
            return true;
        }

        // ── recursive descent ───────────────────────────────────────────
        // expr := term (('+'|'-') term)*
        private static MathExprNode ParseExpr(List<Tok> t, ref int p, string ctx)
        {
            var left = ParseTerm(t, ref p, ctx);
            while (p < t.Count && (t[p].Kind == TokKind.Plus || t[p].Kind == TokKind.Minus))
            {
                var op = t[p].Kind == TokKind.Plus ? '+' : '-'; p++;
                var right = ParseTerm(t, ref p, ctx);
                left = new BinaryNode { Op = op, Left = left, Right = right };
            }
            return left;
        }

        // term := factor (('*'|'/') factor)*
        private static MathExprNode ParseTerm(List<Tok> t, ref int p, string ctx)
        {
            var left = ParseFactor(t, ref p, ctx);
            while (p < t.Count && (t[p].Kind == TokKind.Star || t[p].Kind == TokKind.Slash))
            {
                var op = t[p].Kind == TokKind.Star ? '*' : '/'; p++;
                var right = ParseFactor(t, ref p, ctx);
                left = new BinaryNode { Op = op, Left = left, Right = right };
            }
            return left;
        }

        // factor := '-' factor | '(' expr ')' | number | ref | aggregate
        private static MathExprNode ParseFactor(List<Tok> t, ref int p, string ctx)
        {
            if (p >= t.Count) throw Err(ctx, "expression ended unexpectedly.");
            var tok = t[p];
            switch (tok.Kind)
            {
                case TokKind.Minus:
                    p++; return new UnaryNode { Operand = ParseFactor(t, ref p, ctx) };
                case TokKind.LParen:
                    p++;
                    var inner = ParseExpr(t, ref p, ctx);
                    if (p >= t.Count || t[p].Kind != TokKind.RParen) throw Err(ctx, "expression has an unbalanced '('.");
                    p++; return inner;
                case TokKind.Number:
                    p++; return new NumberNode { Value = tok.Num };
                case TokKind.Ref:
                    p++; return new ColumnRefNode { Node = tok.Node, Column = tok.Col };
                case TokKind.Agg:
                    p++; return new AggregateNode { Func = tok.Func, Node = tok.Node.Value, Column = tok.Col, FilterKey = tok.FilterKey };
                default:
                    throw Err(ctx, "expected a number, field, or '(' in expression.");
            }
        }

        private static InvalidPluginExecutionException Err(string ctx, string msg) =>
            new InvalidPluginExecutionException($"{ctx}: {msg}");
    }
}

File: client/src/editor/model/mathExpr.ts
// Client-side mirror of the engine math-expression parser (Core/Execution/MathExpr.cs):
// numeric literals, {root.<col>} / {node:<guid>.<col>} tokens (same grammar as templateTokens),
// binary + - * /, unary -, and parentheses with standard precedence. Used for author-time
// validation (does it parse?) and to enumerate referenced nodes/columns. Kept in lockstep
// with the engine: same accepted forms, same rejections.

export interface MathRef { node: string | null; column: string; agg?: "sum" | "avg" | "min" | "max" | "count"; filterKey?: string }
export type MathParseResult = { ok: true; refs: MathRef[] } | { ok: false; error: string };

type Tok =
  | { k: "num" } | { k: "ref"; node: string | null; column: string }
  | { k: "agg"; node: string; column: string; func: MathRef["agg"]; filterKey?: string }
  | { k: "+" } | { k: "-" } | { k: "*" } | { k: "/" } | { k: "(" } | { k: ")" };

const GUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
const FILTER_KEY_RE = /^[A-Za-z0-9_]+$/;

function parseToken(token: string): MathRef | string {
  if (token.startsWith("root.")) {
    const column = token.slice("root.".length);
    return column ? { node: null, column } : `token '{${token}}' is missing a column name.`;
  }
  if (token.startsWith("node:")) {
    const rest = token.slice("node:".length);
    const dot = rest.indexOf(".");
    if (dot <= 0 || dot === rest.length - 1) return `token '{${token}}' must be '{node:<id>.<column>}'.`;
    const node = rest.slice(0, dot);
    if (!GUID_RE.test(node) && !node.startsWith("new-")) return `token node '${node}' is not a valid id.`;
    return { node, column: rest.slice(dot + 1) };
  }
  return `unknown token '{${token}}'.`;
}

function parseAggArg(arg: string, func: MathRef["agg"]): { node: string; column: string; filterKey?: string } | string {
  let filterKey: string | undefined;
  const ws = arg.search(/[ \t]/);
  if (ws >= 0) {
    const tail = arg.slice(ws + 1).trim();
    arg = arg.slice(0, ws);
    if (!tail.startsWith("filter:")) return `${func}(...) has unexpected text '${tail}'; expected 'filter:<key>'.`;
    filterKey = tail.slice("filter:".length).trim();
    if (!filterKey || !FILTER_KEY_RE.test(filterKey)) return `${func}(...) filter key '${filterKey}' must be one or more letters, digits, or underscores.`;
  }
  if (!arg.startsWith("node:")) return `${func}(...) argument must be a node reference.`;
  const rest = arg.slice("node:".length);
  const dot = rest.indexOf(".");
  const node = dot < 0 ? rest : rest.slice(0, dot);
  const column = dot < 0 ? "" : rest.slice(dot + 1);
  if (!GUID_RE.test(node) && !node.startsWith("new-")) return `${func}(...) node '${node}' is not a valid id.`;
  if (func === "count") { if (column) return "count(...) takes no column."; }
  else if (!column) return `${func}(...) needs a column.`;
  return filterKey ? { node, column, filterKey } : { node, column };
}

function tokenize(s: string): Tok[] | string {
  const toks: Tok[] = [];
  let i = 0;
  while (i < s.length) {
    const c = s[i];
    if (/\s/.test(c)) { i++; continue; }
    if (c === "+" || c === "-" || c === "*" || c === "/" || c === "(" || c === ")") {
      toks.push({ k: c } as Tok); i++; continue;
    }
    if (c === "{") {
      const close = s.indexOf("}", i + 1);
      if (close < 0) return "expression has an unclosed '{' token.";
      const parsed = parseToken(s.slice(i + 1, close));
      if (typeof parsed === "string") return parsed;
      toks.push({ k: "ref", node: parsed.node, column: parsed.column });
      i = close + 1; continue;
    }
    if (/[0-9.]/.test(c)) {
      let j = i;
      while (j < s.length && /[0-9.]/.test(s[j])) j++;
      const lit = s.slice(i, j);
      if (!/^\d+\.?\d*$|^\.\d+$/.test(lit)) return `'${lit}' is not a valid number.`;
      toks.push({ k: "num" }); i = j; continue;
    }
    if (/[a-zA-Z]/.test(c)) {
      let j = i;
      while (j < s.length && /[a-zA-Z]/.test(s[j])) j++;
      const name = s.slice(i, j).toLowerCase();
      if (!["sum", "avg", "min", "max", "count"].includes(name)) return `unknown function '${name}'.`;
      let k = j;
      while (k < s.length && /\s/.test(s[k])) k++;
      if (s[k] !== "(") return `'${name}' must be followed by '('.`;
      const close = s.indexOf(")", k + 1);
      if (close < 0) return `'${name}(' has no closing ')'.`;
      const arg = s.slice(k + 1, close).trim();
      const parsed = parseAggArg(arg, name as MathRef["agg"]);
      if (typeof parsed === "string") return parsed;
      toks.push({ k: "agg", node: parsed.node, column: parsed.column, func: name as MathRef["agg"], filterKey: parsed.filterKey });
      i = close + 1; continue;
    }
    return `unexpected character '${c}'.`;
  }
  return toks;
}

// Recursive descent mirroring the engine grammar; collects refs as a side effect.
export function parseMathExpr(expr: string): MathParseResult {
  const toks = tokenize(expr ?? "");
  if (typeof toks === "string") return { ok: false, error: toks };
  const refs: MathRef[] = [];
  let p = 0;

  const factor = (): string | null => {
    if (p >= toks.length) return "expression ended unexpectedly.";
    const t = toks[p];
    if (t.k === "-") { p++; return factor(); }
    if (t.k === "(") {
      p++;
      const err = expr2();
      if (err) return err;
      if (p >= toks.length || toks[p].k !== ")") return "expression has an unbalanced '('.";
      p++; return null;
    }
    if (t.k === "num") { p++; return null; }
    if (t.k === "ref") { refs.push({ node: t.node, column: t.column }); p++; return null; }
    if (t.k === "agg") {
      const ref: MathRef = { node: t.node, column: t.column, agg: t.func };
      if (t.filterKey) ref.filterKey = t.filterKey;
      refs.push(ref); p++; return null;
    }
    return "expected a number, field, or '(' in expression.";
  };
  const term = (): string | null => {
    let err = factor(); if (err) return err;
    while (p < toks.length && (toks[p].k === "*" || toks[p].k === "/")) { p++; err = factor(); if (err) return err; }
    return null;
  };
  const expr2 = (): string | null => {
    let err = term(); if (err) return err;
    while (p < toks.length && (toks[p].k === "+" || toks[p].k === "-")) { p++; err = term(); if (err) return err; }
    return null;
  };

  const err = expr2();
  if (err) return { ok: false, error: err };
  if (p !== toks.length) return { ok: false, error: "unexpected trailing input in expression." };
  return { ok: true, refs };
}

// Renders `expr` with every `{root.<col>}` / `{node:<id>.<col>}` token and every aggregate
// call (`sum(node:<id>.<col>)`, `count(node:<id>)`, ...) replaced by `labelFor`'s friendly
// text, leaving operators/numbers/whitespace/parens untouched. Mirrors friendlyTemplate's
// approach: the expression must already parse (grammar-valid), so this rescans it with the
// same per-token parsers instead of re-validating.
export function friendlyMathExpr(
  expr: string,
  labelFor: (ref: MathRef) => string,
  filterSummary?: (key: string) => string
): string {
  if (!parseMathExpr(expr).ok) return expr;
  let out = "";
  let i = 0;
  while (i < expr.length) {
    const c = expr[i];
    if (c === "{") {
      const close = expr.indexOf("}", i + 1);
      const parsed = parseToken(expr.slice(i + 1, close)) as MathRef;
      out += labelFor({ node: parsed.node, column: parsed.column });
      i = close + 1;
      continue;
    }
    if (/[a-zA-Z]/.test(c)) {
      let j = i;
      while (j < expr.length && /[a-zA-Z]/.test(expr[j])) j++;
      const name = expr.slice(i, j).toLowerCase();
      let k = j;
      while (k < expr.length && /\s/.test(expr[k])) k++;
      const close = expr.indexOf(")", k + 1);
      const arg = expr.slice(k + 1, close).trim();
      const parsedArg = parseAggArg(arg, name as MathRef["agg"]) as { node: string; column: string; filterKey?: string };
      const ref: MathRef = { node: parsedArg.node, column: parsedArg.column, agg: name as MathRef["agg"] };
      if (parsedArg.filterKey) ref.filterKey = parsedArg.filterKey;
      let label = labelFor(ref);
      if (ref.filterKey && filterSummary) label += ` where ${filterSummary(ref.filterKey)}`;
      out += label;
      i = close + 1;
      continue;
    }
    out += c;
    i += 1;
  }
  return out;
}
